packages/ai/core/util/process-chat-response-v2.ts

```diff
@@ -0,0 +1,243 @@
+import { LanguageModelV2FinishReason } from '@ai-sdk/provider';
+import { generateId as generateIdFunction } from '@ai-sdk/provider-utils';
+import type { JSONValue, UIMessage, UseChatOptions } from '../types';
+import {
+  calculateLanguageModelUsage,
+  LanguageModelUsage,
+} from '../types/duplicated/usage';
+import { ChatStore } from './chat-store';
+import { processDataStream } from './process-data-stream';
+
+type LocalState = {
+  messageAnnotations: JSONValue[];
+  usage: LanguageModelUsage;
+  finishReason: LanguageModelV2FinishReason;
+};
+
+export async function processChatResponseV2({
+  chatId,
+  stream,
+  update,
+  onToolCall,
+  onFinish,
+  generateId = generateIdFunction,
+  store,
+}: {
+  chatId: string;
+  stream: ReadableStream<Uint8Array>;
+  update: (options: { data: JSONValue[] | undefined }) => void;
+  onToolCall?: UseChatOptions['onToolCall'];
+  onFinish?: (options: {
+    message: UIMessage | undefined;
+    finishReason: LanguageModelV2FinishReason;
+    usage: LanguageModelUsage;
+  }) => void;
+  generateId?: () => string;
+  store: ChatStore;
+}) {
+  const data: JSONValue[] = [];
+
+  const lastMessage = store.getLastMessage(chatId);
+  const replaceLastMessage = lastMessage?.role === 'assistant';
+
+  const localState: LocalState = {
+    // We keep track of message annotation locally because they may be returned before the active response is created (before native parts are processed):
+    messageAnnotations: replaceLastMessage
+      ? (lastMessage?.annotations ?? [])
+      : [],
+    usage: {
+      completionTokens: NaN,
+      promptTokens: NaN,
+      totalTokens: NaN,
+    },
+    finishReason: 'unknown',
+  };
+
+  const updateMessageStore = (partDelta: UIMessage['parts'][number]) => {
+    update({ data: [] }); // TODO: Sets status to streaming (SWR), we should separate data updates from status updates
```

Comment:
i wonder if this could also be simplified, e.g. i'm considering removing `data` fully in ai sdk 5 (only data parts on assistant messages) and waiting to see if people scream